<a href="https://colab.research.google.com/github/IvankoPilotOfHalychyna/Foundations_AI/blob/main/Tutorial9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tutorial 9

## Introduction

### Naive Bayes Algorithm

The Naive Bayes algorithm is a classic probabilistic algorithm used primarily for classification. It is a model based on Bayes’ theorem with the assumption of feature independence. This means that for a given classified object, individual features (variables) are treated as independent, which significantly simplifies the computations.

## Bayes' Theorem

Bayes’ theorem describes the relationship between conditional, prior, and posterior probabilities:

$$
P(C \mid X) = \frac{P(X \mid C) P(C)}{P(X)}
$$

where:

- $ P(C \mid X) $ — the probability that an object belongs to class $ C $, given features $ X $ (posterior),
- $ P(X \mid C) $ — the probability of observing features $ X $ if the object belongs to class $ C $ (likelihood),
- $ P(C) $ — the prior probability that an object belongs to class $ C $,
- $ P(X) $ — the probability of observing features $ X $ (often treated as a constant, ignored in classification).

<br>

## Assumptions of Naive Bayes

Naive Bayes assumes that the features are independent. This means that each feature is treated separately in the calculations, ignoring mutual dependencies between them. Thanks to this, the algorithm is very fast and efficient, although it does not always achieve the best results compared to more advanced models.

## Modeling in Naive Bayes

To compute the probability that an object belongs to class $ C $, we use the following formula:

$$
P(C \mid X) = P(C) \prod_{i=1}^{n} P(x_i \mid C) \frac{1}{P(X)}
$$

$$
\prod_{i=1}^{n} P(x_i \mid C) = P(x_1 \mid C) \cdot P(x_2 \mid C) \cdot \cdots \cdot P(x_i \mid C)
$$

$$
P(C \mid X) = \frac {P(C) \cdot P(x_1 \mid C) \cdot P(x_2 \mid C) \cdot \cdots \cdot P(x_i \mid C)}{P(X)}
$$

<br>

We may also ignore the denominator, because it is the same for all classes and does not affect the classification. Then:

$$
P(C \mid X) \propto P(C) \cdot P(x_1 \mid C) \cdot P(x_2 \mid C) \cdot \cdots \cdot P(x_i \mid C)
$$

where:

- $ x_1, x_2, \dots, x_n $ are the features of the object $ X $,
- $ P(x_i \mid C) $ is the probability of observing feature $ x_i $ given that the object belongs to class $ C $.

In practice, the algorithm computes probabilities for each class and selects the class with the highest probability.

### Step by step:

1. Compute the prior probability $ P(C) $ for each class $ C $, i.e., the probability that a given object belongs to class $ C $:

$$
P(C) = \frac{\text{number of samples in class } C}{\text{total number of samples}}
$$

2. Compute the conditional probabilities $ P(x_i \mid C) $ for each feature $ x_i $ and class $ C $. This can be done based on the frequency of the feature $ x_i $ in class $ C $.

3. Apply Bayes’ formula for each class $ C $ and choose the class for which $ P(C \mid X) $ is the highest.


### Example of a Computational Task Using the Bayes Classifier
Below is a dataset presenting object attributes and their classification.

| Feature 1 (x1) | Feature 2 (x2) | Class (C) |
|----------------|----------------|-----------|
| Yes            | Yes            | A         |
| Yes            | No             | A         |
| No             | Yes            | B         |
| No             | Yes            | A         |
| Yes            | No             | B         |

<br>

### Step 1: Compute the prior probabilities $P(A)$ and $P(B)$.

$$
P(A) = \frac{\text{number of samples in class A}}{\text{total number of samples}}
$$

$$
P(B) = \frac{\text{number of samples in class B}}{\text{total number of samples}}
$$

<br>

### Step 2: Compute the conditional probabilities $P(x_1 \mid A)$, $P(x_1 \mid B)$, $P(x_2 \mid A)$, and $P(x_2 \mid B)$.

For each feature, compute how frequently it appears in each class.

<br>

### Step 3: Based on the calculated probabilities, compute $P(A \mid X)$ and $P(B \mid X)$, assuming we have an object with feature 1 = "Yes" and feature 2 = "No".

$$ X = [Yes, No] $$

The task is to determine to which class the new object will belong.


### Calculating Probabilities Using Bayes' Formula

## 1. Compute the prior probabilities $ P(A) $ and $ P(B) $

We have 5 samples in the dataset, of which:
- 3 samples belong to class $ A $ (in the first, second, and fourth rows),
- 2 samples belong to class $ B $ (in the third and fifth rows).

We compute the prior probabilities for each class:

$$
P(A) = \frac{3}{5} = 0.6
$$

$$
P(B) = \frac{2}{5} = 0.4
$$

<br>

## 2. Compute the conditional probabilities $ P(x_1 \mid A) $, $ P(x_1 \mid B) $, $ P(x_2 \mid A) $, and $ P(x_2 \mid B) $

### Feature 1 ($ x_1 $):

$$
P(x_1 = \text{Yes} \mid A) \quad \text{— How many times is feature 1 equal to "Yes" in class } A?
$$

In class $ A $ we have 3 samples, of which 2 have feature 1 equal to "Yes" (the first and second samples). Thus:

$$
P(x_1 = \text{Yes} \mid A) = \frac{2}{3}
$$

<br>

$$
P(x_1 = \text{Yes} \mid B) \quad \text{— How many times is feature 1 equal to "Yes" in class } B?
$$

In class $ B $ we have 2 samples, of which 1 has feature 1 equal to "Yes" (the third sample). Thus:

$$
P(x_1 = \text{Yes} \mid B) = \frac{1}{2}
$$

<br>
<br>

### Feature 2 ($ x_2 $):

$$
P(x_2 = \text{Yes} \mid A) \quad \text{— How many times is feature 2 equal to "Yes" in class } A?
$$

In class $ A $ we have 3 samples, of which 2 have feature 2 equal to "Yes" (the first and fourth samples). Thus:

$$
P(x_2 = \text{Yes} \mid A) = \frac{2}{3}
$$

$$
P(x_2 = \text{Yes} \mid B) \quad \text{— How many times is feature 2 equal to "Yes" in class } B?
$$

In class $ B $ we have 2 samples, of which 1 has feature 2 equal to "Yes" (the third sample). Thus:

$$
P(x_2 = \text{Yes} \mid B) = \frac{1}{2}
$$

<br>
<br>

## 3. Compute $ P(A \mid X) $ and $ P(B \mid X) $

We have a new object with the following features:
- Feature 1 = "Yes"
- Feature 2 = "No"

Thus we compute $ P(x_1 = \text{Yes}, x_2 = \text{No}) $.

### Calculating $ P(A \mid X) $:

Let us use Bayes’ formula for class $ A $:

$$
P(A \mid X) \propto P(A) \cdot P(x_1 = \text{Yes} \mid A) \cdot P(x_2 = \text{No} \mid A)
$$

$$
P(x_1 = \text{Yes} \mid A) = \frac{2}{3}, \quad P(x_2 = \text{No} \mid A) = 1 - \frac{2}{3} = \frac{1}{3}
$$

$$
P(A \mid X) \propto 0.6 \cdot \frac{2}{3} \cdot \frac{1}{3} = 0.6 \cdot \frac{2}{9} = \frac{1.2}{9} = \frac{2}{15}
$$

### Calculating $ P(B \mid X) $:

Use Bayes’ formula for class $ B $:

$$
P(B \mid X) \propto P(B) \cdot P(x_1 = \text{Yes} \mid B) \cdot P(x_2 = \text{No} \mid B)
$$

$$
P(x_1 = \text{Yes} \mid B) = \frac{1}{2}, \quad P(x_2 = \text{No} \mid B) = 1 - \frac{1}{2} = \frac{1}{2}
$$

$$
P(B \mid X) \propto 0.4 \cdot \frac{1}{2} \cdot \frac{1}{2} = 0.4 \cdot \frac{1}{4} = \frac{0.4}{4} = \frac{1}{10}
$$

## 4. Class Selection

Comparing the values of $ P(A \mid X) $ and $ P(B \mid X) $, we get:

$$
P(A \mid X) = \frac{2}{15}, \quad P(B \mid X) = \frac{1}{10} = \frac{1.5}{15}
$$

Since $ P(B \mid X) > P(A \mid X) $, the object will be assigned to class $ B $.


### Example of Using the Bayes Classifier in Python Code

In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, classification_report

iris = load_iris()
X = iris.data
y = iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

model = GaussianNB()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.4f}")
print(classification_report(y_test, y_pred))


Accuracy: 0.9778
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      0.92      0.96        13
           2       0.93      1.00      0.96        13

    accuracy                           0.98        45
   macro avg       0.98      0.97      0.97        45
weighted avg       0.98      0.98      0.98        45



## Task 1
For the table below, classify the vector $ X = [No, Yes] $ using the Naive Bayes classifier.

| Feature 1 (x1) | Feature 2 (x2) | Class (C) |
|----------------|----------------|-----------|
| No             | Yes            | C         |
| Yes            | No             | A         |
| No             | Yes            | D         |
| Yes            | No             | A         |
| No             | No             | D         |
| No             | Yes            | C         |


# Task 1 solution

In [1]:
import pandas as pd

# Data provided in the table
data = {
    'Feature 1': ['No', 'Yes', 'No', 'Yes', 'No', 'No'],
    'Feature 2': ['Yes', 'No', 'Yes', 'No', 'No', 'Yes'],
    'Class': ['C', 'A', 'D', 'A', 'D', 'C']
}
df = pd.DataFrame(data)

print("Original Dataset:")
print(df)
print("\n")

# New vector to classify
X_new = {'Feature 1': 'No', 'Feature 2': 'Yes'}

# 1. Calculate Prior Probabilities P(C)
class_counts = df['Class'].value_counts()
total_samples = len(df)
prior_probabilities = class_counts / total_samples

print("Prior Probabilities (P(C)):")
print(prior_probabilities)
print("\n")

# 2. Calculate Conditional Probabilities P(Feature_value | Class)
conditional_probabilities = {}
for feature in ['Feature 1', 'Feature 2']:
    conditional_probabilities[feature] = {}
    for cls in df['Class'].unique():
        df_class = df[df['Class'] == cls]
        feature_value_counts = df_class[feature].value_counts()
        # Add Laplace smoothing to handle zero probabilities, although for this specific dataset it won't change the result due to zero P(A|X) and P(D|X).
        # Adding 1 to numerator and number of unique feature values to denominator.
        # This makes it robust to cases where a feature value doesn't appear in a class.
        # For this specific task, let's keep it without smoothing first to match manual calculation exactly, then mention smoothing importance.

        # Using direct counts for this problem as per manual calculation example structure.
        cond_probs = feature_value_counts / len(df_class)
        conditional_probabilities[feature][cls] = cond_probs.to_dict()

print("Conditional Probabilities (P(Feature_value | Class)):")
for feature, classes in conditional_probabilities.items():
    print(f"  {feature}:")
    for cls, probs in classes.items():
        print(f"    Class {cls}: {probs}")
print("\n")

# 3. Classify the new vector X = [No, Yes]
posterior_scores = {}
for cls in df['Class'].unique():
    score = prior_probabilities[cls]

    # P(x1=No | Class)
    feature1_val = X_new['Feature 1']
    prob_feature1_given_class = conditional_probabilities['Feature 1'][cls].get(feature1_val, 0) # Use .get with 0 for unseen values

    # P(x2=Yes | Class)
    feature2_val = X_new['Feature 2']
    prob_feature2_given_class = conditional_probabilities['Feature 2'][cls].get(feature2_val, 0) # Use .get with 0 for unseen values

    score *= prob_feature1_given_class * prob_feature2_given_class
    posterior_scores[cls] = score

print(f"New vector X: {X_new}")
print("Posterior Scores (P(C) * P(x1 | C) * P(x2 | C)):")
print(posterior_scores)

# Select the class with the highest posterior score
predicted_class = max(posterior_scores, key=posterior_scores.get)

print(f"\nThe classified class for X = {X_new} is: {predicted_class}")


Original Dataset:
  Feature 1 Feature 2 Class
0        No       Yes     C
1       Yes        No     A
2        No       Yes     D
3       Yes        No     A
4        No        No     D
5        No       Yes     C


Prior Probabilities (P(C)):
Class
C    0.333333
A    0.333333
D    0.333333
Name: count, dtype: float64


Conditional Probabilities (P(Feature_value | Class)):
  Feature 1:
    Class C: {'No': 1.0}
    Class A: {'Yes': 1.0}
    Class D: {'No': 1.0}
  Feature 2:
    Class C: {'Yes': 1.0}
    Class A: {'No': 1.0}
    Class D: {'Yes': 0.5, 'No': 0.5}


New vector X: {'Feature 1': 'No', 'Feature 2': 'Yes'}
Posterior Scores (P(C) * P(x1 | C) * P(x2 | C)):
{'C': np.float64(0.3333333333333333), 'A': np.float64(0.0), 'D': np.float64(0.16666666666666666)}

The classified class for X = {'Feature 1': 'No', 'Feature 2': 'Yes'} is: C


## Task 2
For the `load_breast_cancer` dataset from the `sklearn.datasets` module, perform DEA analysis and classification using the `Naive Bayes` classifier. Remember to scale the data.

<br>

Example of loading the data:

```
from sklearn.datasets import load_breast_cancer

data = load_breast_cancer()
X = data.data
y = data.target
```

# Task 2 solution

In [2]:
from sklearn.datasets import load_breast_cancer

# Load the breast_cancer dataset
data = load_breast_cancer()
X = data.data
y = data.target

print("Breast Cancer dataset loaded successfully.")
print(f"Shape of features (X): {X.shape}")
print(f"Shape of target (y): {y.shape}")

Breast Cancer dataset loaded successfully.
Shape of features (X): (569, 30)
Shape of target (y): (569,)


## Exploratory Data Analysis (EDA)


In [3]:
import pandas as pd

# 1. Convert the feature data X into a pandas DataFrame
df_X = pd.DataFrame(X, columns=data.feature_names)

print("### Feature DataFrame (df_X) ###\n")

# 2. Print the first 5 rows of df_X
print("First 5 rows of df_X:")
print(df_X.head())
print("\n")

# 3. Display a concise summary of the df_X DataFrame
print("Concise summary of df_X:")
df_X.info()
print("\n")

# 4. Generate descriptive statistics for df_X
print("Descriptive statistics for df_X:")
print(df_X.describe())
print("\n")

# 5. Print the target names
print("Target names:")
print(data.target_names)
print("\n")

# 6. Calculate and print the distribution of the target variable y
print("Distribution of target variable (y):")
# Convert y to a Series for easier value counting and use target names for labels
df_y = pd.Series(y).map(lambda x: data.target_names[x])
print(df_y.value_counts())


### Feature DataFrame (df_X) ###

First 5 rows of df_X:
   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  wors

## Scale Data


In [4]:
from sklearn.preprocessing import StandardScaler

# Initialize the StandardScaler
scaler = StandardScaler()

# Scale the features (X)
X_scaled = scaler.fit_transform(X)

print("Features scaled successfully. First 5 rows of scaled data:")
print(X_scaled[:5])

Features scaled successfully. First 5 rows of scaled data:
[[ 1.09706398e+00 -2.07333501e+00  1.26993369e+00  9.84374905e-01
   1.56846633e+00  3.28351467e+00  2.65287398e+00  2.53247522e+00
   2.21751501e+00  2.25574689e+00  2.48973393e+00 -5.65265059e-01
   2.83303087e+00  2.48757756e+00 -2.14001647e-01  1.31686157e+00
   7.24026158e-01  6.60819941e-01  1.14875667e+00  9.07083081e-01
   1.88668963e+00 -1.35929347e+00  2.30360062e+00  2.00123749e+00
   1.30768627e+00  2.61666502e+00  2.10952635e+00  2.29607613e+00
   2.75062224e+00  1.93701461e+00]
 [ 1.82982061e+00 -3.53632408e-01  1.68595471e+00  1.90870825e+00
  -8.26962447e-01 -4.87071673e-01 -2.38458552e-02  5.48144156e-01
   1.39236330e-03 -8.68652457e-01  4.99254601e-01 -8.76243603e-01
   2.63326966e-01  7.42401948e-01 -6.05350847e-01 -6.92926270e-01
  -4.40780058e-01  2.60162067e-01 -8.05450380e-01 -9.94437403e-02
   1.80592744e+00 -3.69203222e-01  1.53512599e+00  1.89048899e+00
  -3.75611957e-01 -4.30444219e-01 -1.46748968e-0

## Split Dataset


In [5]:
from sklearn.model_selection import train_test_split

# Split the scaled data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3, random_state=42)

print("Data split into training and testing sets successfully.")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Data split into training and testing sets successfully.
X_train shape: (398, 30)
X_test shape: (171, 30)
y_train shape: (398,)
y_test shape: (171,)


## Train a Naive Bayes classifier



In [6]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, classification_report

# Initialize the Naive Bayes classifier (Gaussian Naive Bayes is suitable for continuous data)
model = GaussianNB()

# Train the model on the scaled training data
model.fit(X_train, y_train)

# Make predictions on the scaled test data
y_pred = model.predict(X_test)

# Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
print(f"Model Accuracy: {accuracy:.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred, target_names=data.target_names))

Model Accuracy: 0.9357

Classification Report:
              precision    recall  f1-score   support

   malignant       0.92      0.90      0.91        63
      benign       0.94      0.95      0.95       108

    accuracy                           0.94       171
   macro avg       0.93      0.93      0.93       171
weighted avg       0.94      0.94      0.94       171



## Summary:

### Data Analysis Key Findings
*   The `breast_cancer` dataset was successfully loaded, containing 569 samples and 30 features.
*   Exploratory Data Analysis revealed no missing values across the 30 features, and the dataset included both 'malignant' and 'benign' target classes.
*   The target variable distribution showed 357 'benign' cases and 212 'malignant' cases.
*   Features were successfully scaled using `StandardScaler`.
*   The data was split into training (398 samples) and testing (171 samples) sets, with a test size of 30%.
*   A Gaussian Naive Bayes classifier was trained and achieved an accuracy of 0.9357 on the test set.
*   The model demonstrated good performance for both classes, with a precision of 0.92, recall of 0.90, and F1-score of 0.91 for 'malignant' cases, and precision of 0.94, recall of 0.95, and F1-score of 0.95 for 'benign' cases.